### PR TITLE
Modifications for Ruby 1.9 and better

### PR DESCRIPTION
**Pull request by gflohr**
The module compiles and links and passes all tests on these platforms:

- Gentoo Linux, ruby 2.1.9
- Gentoo Linux, ruby 2.2.6
- MacPorts, ruby 1.8.7
- MacPorts, ruby 1.9.3
- MacPorts, ruby 2.0.0
- MacPorts, ruby 2.4.1
- Mac OS 10.12.6 (Sierra), ruby 2.0.0

Changes:

- fixed compiler and linker flags for recent ruby versions
- fixed linker flags for MacPorts
- rewritten namespace inspection (did not work for ruby 1.9 or better)
- fixed various compiler warnings

TODO:

The documentation should be updated.

### CODE REVIEW

**Comment by shlomif** [comment]
Hi Guido!

On Tue, 15 Aug 2017 11:31:39 +0000 (UTC)
Guido Flohr <notifications@github.com> wrote:

> The module compiles and links and passes all tests on these platforms:
> 
> - Gentoo Linux, ruby 2.1.9
> - Gentoo Linux, ruby 2.2.6
> - MacPorts, ruby 1.8.7
> - MacPorts, ruby 1.9.3
> - MacPorts, ruby 2.0.0
> - MacPorts, ruby 2.4.1
> - Mac OS 10.12.6 (Sierra), ruby 2.0.0
> 
> Changes:
> 
> - fixed compiler and linker flags for recent ruby versions
> - fixed linker flags for MacPorts
> - rewritten namespace inspection (did not work for ruby 1.9 or better)
> - fixed various compiler warnings
> 

Sounds very cool! Many thanks! I'd like to take a look at it soon.

Regards,

	Shlomi



-- 
-----------------------------------------------------------------
Shlomi Fish       http://www.shlomifish.org/
http://youtu.be/xZLwtc9x4yA - Anime in Real Life!! (Parody)

Wikipedia deletionists Don’t Die. They lose notability and get deleted.
    — http://www.shlomifish.org/humour.html

Please reply to list if it's a mailing list post - http://shlom.in/reply .

**Comment by shlomif** [comment]
Merged, thanks!

**Comment by gflohr** [comment]
Hi,

> On Aug 16, 2017, at 14:53, Shlomi Fish <notifications@github.com> wrote:
> 
> Merged, thanks!
> 
> 

I thank you! :)

I think, I should elaborate just a little on the central change, though.

The original code contained a Ruby routine inline_ruby_class_grokker() that was first compiled with rb_eval() and then executed, whenever Ruby code was compiled.  All that happened in Ruby.pm and the Ruby code read like this:

def inline_ruby_class_grokker(*classes)
    if classes == []
	ObjectSpace.each_object(Class) do |x|
	    yield ['classes', x.name]
	end
	ObjectSpace.each_object(Module) do |x|
	    yield ['modules', x.name]
	end
	Kernel.methods.each do |x|
	    yield ['functions', x]
	end
    else
	classes.each do |k|
	    n = {}
	    begin
		n['methods'] = eval "#{k}.methods"
		n['imethods'] = eval "#{k}.instance_methods"
	    rescue Exception
		p "Exception: " + $!
	    end
	    yield [k, n]

	end
    end
end

That Ruby code was later executed like this from Perl:

    rb_iter(undef, sub {my ($type, $name) = @_; $pre{$type}{$name}++})
          ->inline_ruby_class_grokker;

Ruby 1.9 (and all later versions) bailed out here with a Ruby exception "LocalJumpError: no block given (yield)”.  To be honest, I didn’t understand the error and I couldn’t debug  the error because Ruby threw the exception and because of that I couldn’t debug the error and therefore I didn’t understand the error and …

I rewrote the Ruby code and split it up in three parts that now get parsed and evaluated whenever new Ruby code is fed into Inline::Ruby.  That is a little bit less efficient but I think that this is negligible for typical use cases.  It hides, however, the fact that obviously rb_iter() no longer works like expected for ruby 1.9+.  That should maybe be investigated further or we should just wait for bug reports if anybody really uses rb_iter() which is not really a standard interface for Inline modules.

In other words: The module most probably is still imperfect but - knock, knock, knock on wood - maybe nobody will notice it or care.

Many greetings from Sofia!
Guido
— 
http://www.guido-flohr.net/
http://www.cantanea.com/careers/ • facebook.com/cantaneacom • twitter.com/cantaneacom
